### PR TITLE
speed up loops using numba

### PR DESCRIPTION
**Pull request by hoffmanlucas**
This version using numba jit runs more than 10 times faster. 

```
PS D:\OneDrive\GitHub Packets\IMD-calculator-1> python .\python\calculator.py 11 101 102 103 104 105 106 107 108 109 110
ORDER: 11
FREQUENCIES: [101.0, 102.0, 103.0, 104.0, 105.0, 106.0, 107.0, 108.0, 109.0, 110.0]
Time taken: 5.56754 seconds
Found 5188581 IMD product
```s

### CODE REVIEW

File: README.md
Comment (JolonB) [inline README.md:1]: Could you update the `requirements.txt` file to include `numba`?

Something like `numba~=0.50` would be good (strangely, my pip won't install the latest version, so I've dropped it to 0.50 or later)

**Comment by JolonB** [comment]
This is great! I tried learning to use numba when I first made this but I couldn't get it to work. When I run it on my PC (Ubuntu 20, Python 3.8.10, numba 0.58.1), I get the following warnings:

```
$ python calculator.py 11 101 102 103 104 105 106 107 108 109 110
calculator.py:55: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.
  def add_positive_signal(
calculator.py:107: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.
  def add_negative_signal(
ORDER: 11
FREQUENCIES: [101.0, 102.0, 103.0, 104.0, 105.0, 106.0, 107.0, 108.0, 109.0, 110.0]
calculator.py:79: NumbaPendingDeprecationWarning: 
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'products' of function 'add_negative_signal'.

For more information visit https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "calculator.py", line 107:
@numba.jit
def add_negative_signal(
^

  add_positive_signal(
calculator.py:79: NumbaPendingDeprecationWarning: 
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'products' of function 'add_positive_signal'.

For more information visit https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "calculator.py", line 55:
@numba.jit
def add_positive_signal(
^

  add_positive_signal(
calculator.py:79: NumbaPendingDeprecationWarning: 
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'transmit_freqs' of function 'add_negative_signal'.

For more information visit https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "calculator.py", line 107:
@numba.jit
def add_negative_signal(
^

  add_positive_signal(
calculator.py:79: NumbaPendingDeprecationWarning: 
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'transmit_freqs' of function 'add_positive_signal'.

For more information visit https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "calculator.py", line 55:
@numba.jit
def add_positive_signal(
^

  add_positive_signal(
/home/jolon/IMD-calculator/python/venv/lib/python3.8/site-packages/numba/core/ir_utils.py:2172: NumbaPendingDeprecationWarning: 
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'products' of function 'add_positive_signal'.

For more information visit https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "calculator.py", line 55:
@numba.jit
def add_positive_signal(
^

  warnings.warn(NumbaPendingDeprecationWarning(msg, loc=loc))
/home/jolon/IMD-calculator/python/venv/lib/python3.8/site-packages/numba/core/ir_utils.py:2172: NumbaPendingDeprecationWarning: 
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'transmit_freqs' of function 'add_positive_signal'.

For more information visit https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "calculator.py", line 55:
@numba.jit
def add_positive_signal(
^

  warnings.warn(NumbaPendingDeprecationWarning(msg, loc=loc))
Time taken: 5.38905 seconds
Found 5188581 IMD products
```

Any idea what might be causing these?

File: python/calculator.py [hunk 1]
```diff
@@ -28,53 +29,90 @@ def calculate(self, order: int):
         elif order < 3:
             raise ValueError("order must be 3 or higher. Got {}".format(order))
         self.order = order
-        self.products = list()
+        
+        # pass in the first product so NUMBA can infer the types
+        # this is a hacky way to get around the fact that NUMBA doesn't support lists
+        # Remove it from the final result
```
Comment (JolonB) [inline python/calculator.py:35]: I have a feeling this hasn't been removed, has it?
When I run the original code, I get 1 fewer IMD product than running the numba code:

```
~/IMD-calculator/python$ python calculator.py 11 101 102 103 104 105 106 107 108 109 110  # original code
ORDER: 11
FREQUENCIES: [101.0, 102.0, 103.0, 104.0, 105.0, 106.0, 107.0, 108.0, 109.0, 110.0]
Time taken: 36.50073 seconds
Found 5188580 IMD products
```

```
~/IMD-calculator-1/python$ python calculator.py 11 101 102 103 104 105 106 107 108 109 110  # new code
... warnings ...
ORDER: 11
FREQUENCIES: [101.0, 102.0, 103.0, 104.0, 105.0, 106.0, 107.0, 108.0, 109.0, 110.0]
... warnings ...
Time taken: 5.24376 seconds
Found 5188581 IMD products
```

**Review by JolonB — CHANGES_REQUESTED** [review]

File: python/calculator.py (hunk 1 above)
Comment (hoffmanlucas) [reply python/calculator.py:35]: you are right. I have pulled a version that removes the first item from the result.

File: README.md
Comment (hoffmanlucas) [reply README.md:1]: done. I am using python 3.12 and numba 0.59.1.

**Comment by hoffmanlucas** [comment]
the issue with your env might be solved by adding the parameter nopython=True to the numba.jit, like 

`@numba.jit(nopython=True)`

It seems be related to an older numba version that requires the parameter explicitly.

**Review by JolonB — APPROVED** [review]

**Comment by JolonB** [comment]
@hoffmanlucas Thanks for this update!

**Pull request merged**